From: Kamil Bienkiewicz <perceivalpercy@gmail.com>
Subject: [PATCH] AP MLD: Insert PMKR1Name into per-link RSNEs in EAPOL-Key msg 3/4

For a non-AP MLD, EAPOL-Key msg 3/4 does not carry a common RSNE; the
RSNE of each affiliated AP is delivered in the MLO Link KDEs instead.
The FT initial mobility domain association code still tried to insert
the PMKR1Name into the (missing) common RSNE, wpa_insert_pmkid() failed
on the empty buffer, and the 4-way handshake was aborted with "FT:
Failed to insert PMKR1Name into RSN IE in EAPOL-Key data" for every
non-AP MLD using an FT AKM.

Skip the common RSNE handling for MLD and insert the PMKR1Name into the
RSNE of each link in the MLO Link KDE, reserving room for it in the KDE
length calculation.

Signed-off-by: Kamil Bienkiewicz <perceivalpercy@gmail.com>
---
--- a/src/ap/wpa_auth.c
+++ b/src/ap/wpa_auth.c
@@ -4744,8 +4744,14 @@
 
 		ie = get_ie(wpa_auth->wpa_ie, wpa_auth->wpa_ie_len,
 			    WLAN_EID_RSN);
-		if (ie)
+		if (ie) {
 			kde_len += 2 + ie[1];
+#ifdef CONFIG_IEEE80211R_AP
+			/* PMKR1Name may be added into the RSNE */
+			if (wpa_key_mgmt_ft(sm->wpa_key_mgmt))
+				kde_len += 2 + PMKID_LEN;
+#endif /* CONFIG_IEEE80211R_AP */
+		}
 
 		ie = get_ie(wpa_auth->wpa_ie, wpa_auth->wpa_ie_len,
 			    WLAN_EID_RSNX);
@@ -4799,6 +4805,9 @@
 		const u8 *rsne, *rsnxe, *rsnoe, *rsno2e, *rsnxoe;
 		size_t rsne_len, rsnxe_len, rsnoe_len, rsno2e_len, rsnxoe_len;
 		size_t kde_len;
+#ifdef CONFIG_IEEE80211R_AP
+		u8 rsne_buf[2 + 255];
+#endif /* CONFIG_IEEE80211R_AP */
 
 		wpa_auth = wpa_get_link_auth(sm->wpa_auth, link_id);
 		if (!wpa_auth)
@@ -4808,6 +4817,29 @@
 			     WLAN_EID_RSN);
 		rsne_len = rsne ? 2 + rsne[1] : 0;
 
+#ifdef CONFIG_IEEE80211R_AP
+		if (rsne && wpa_key_mgmt_ft(sm->wpa_key_mgmt)) {
+			/* FT initial mobility domain association: the RSNE of
+			 * each affiliated AP in EAPOL-Key msg 3/4 includes the
+			 * PMKR1Name (IEEE Std 802.11be, 12.7.6.4). */
+			if (rsne_len + 2 + PMKID_LEN > sizeof(rsne_buf)) {
+				wpa_printf(MSG_ERROR,
+					   "RSN: MLO Link: link=%u RSNE too long for PMKR1Name",
+					   link_id);
+				return NULL;
+			}
+			os_memcpy(rsne_buf, rsne, rsne_len);
+			if (wpa_insert_pmkid(rsne_buf, &rsne_len,
+					     sm->pmk_r1_name, true) < 0) {
+				wpa_printf(MSG_ERROR,
+					   "FT: Failed to insert PMKR1Name into RSNE for MLO link %u",
+					   link_id);
+				return NULL;
+			}
+			rsne = rsne_buf;
+		}
+#endif /* CONFIG_IEEE80211R_AP */
+
 		rsnxe = get_ie(wpa_auth->wpa_ie, wpa_auth->wpa_ie_len,
 			       WLAN_EID_RSNX);
 		rsnxe_len = rsnxe ? 2 + rsnxe[1] : 0;
@@ -5246,7 +5278,9 @@
 		pos += wpa_ie_len;
 	}
 #ifdef CONFIG_IEEE80211R_AP
-	if (wpa_key_mgmt_ft(sm->wpa_key_mgmt)) {
+	/* For a non-AP MLD, the RSNEs are delivered in the MLO Link KDEs and
+	 * the PMKR1Name is inserted into each of them there. */
+	if (wpa_key_mgmt_ft(sm->wpa_key_mgmt) && !is_mld) {
 		int res;
 		size_t elen;
 
